python/ray/train/torch.py

```diff
@@ -316,14 +316,16 @@ def prepare_data_loader(
         def with_sampler(loader):
             # Automatically set the DistributedSampler
 
-            # If using a sampler, the shuffle attribute in the
-            # DataLoader must be set to False.
-            # Instead the shuffling is determined by the shuffle attribute
-            # in the DistributedSampler.
-            # We identify if shuffling is enabled in the passed in
-            # DataLoader by seeing if the sampler for the DataLoader is a
-            # SequentialSampler.
-            shuffle = not isinstance(loader.sampler, SequentialSampler)
+            # If you're using a sampler, the DataLoader shuffle flag must be set to
+            # False. Shuffling is instead determined by the shuffle argument passed to
+            # the DistributedSampler constructor.
+
+            # If no sampler is passed to the DataLoader constructor, Torch constructs a
+            # default sampler. The default sampler is a RandomSampler if shuffling is
+            # enabled and a SequentialSampler otherwise. There is no DataLoader.shuffle
+            # attribute. So, we identify whether shuffling is enabled by checking the
+            # default sampler type.
+            shuffle = isinstance(loader.sampler, RandomSampler)
```

Comment:
Can we log a message when the Sampler is neither `RandomSampler` nor `SequentialSampler`, and indicate to the user what the default behavior is?

Reply:
What do you mean by indicating the default behavior to the user?

It should always be one of the two, so we can probably `assert isinstance(loader.sampler, (RandomSampler, SequentialSampler))`

Reply:
Ah but if the original `DataLoader` can be initialized with a `sampler` instead of the `shuffle` flag, right?

Reply:
Yep lets log this, saying a sampler has been already set, but overriding with DistributedSampler. You can disable this by setting `with_sampler` to False in `prepare_data_loader`.

Reply:
Also, I believe the reason we had `shuffle = not isinstance(loader.sampler, SequentialSampler)` before is because there are multiple types of random samplers (for example `WeightedRandomSampler`), so we just treat all those as `shuffle=True` case.